File: app/src/main/java/lt/markmerkk/utils/UserSettingsImpl.java
package lt.markmerkk.utils;

import lt.markmerkk.JiraSearchJQL;
import lt.markmerkk.listeners.WorldEvents;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;

/**
 * Created by mariusmerkevicius on 12/21/15.
 * Controller for holding persistent data
 */
public class UserSettingsImpl implements UserSettings, WorldEvents {
    public static final String HOST = "HOST";
    public static final String USER = "USER";
    public static final String PASS = "PASS";
    public static final String VERSION = "VERSION";
    public static final String ISSUE_JQL = "ISSUE_JQL";

    AdvHashSettings settings;

    // Used settings
    String host;
    String username;
    String password;
    String issueJql;
    int version = -1;

    public UserSettingsImpl() {
        settings = new AdvHashSettings();
    }

    //region Getters / Setters

    @Override
    public String getIssueJql() {
        if (issueJql == null) return JiraSearchJQL.DEFAULT_JQL_USER_ISSUES;
        return issueJql;
    }

    @Override
    public void setIssueJql(String issueJql) {
        this.issueJql = issueJql;
        settings.save();
    }

    @Override
    public String getHost() {
        if (host == null) return "";
        return host;
    }

    @Override
    public void setHost(String host) {
        this.host = host;
    }

    @Override
    public String getUsername() {
        if (username == null) return "";
        return username;
    }

    @Override
    public void setUsername(String username) {
        this.username = username;
    }

    @Override
    public String getPassword() {
        if (password == null) return "";
        return password;
    }

    @Override
    public void setPassword(String password) {
        this.password = password;
    }

    @Override
    public void setCustom(String key, String value) {
        settings.set(key, value);
        settings.save();
    }

    @Override
    public String getCustom(String key) {
        return settings.get(key);
    }

    @Override
    public int getVersion() {
        return version;
    }

    @Override
    public void setVersion(int version) {
        this.version = version;
        settings.save();
    }

    //endregion

    //region World events

    @PostConstruct
    @Override
    public void onStart() {
        settings.load();
        host = settings.get(HOST);
        username = settings.get(USER);
        password = settings.get(PASS);
        String versionString = settings.get(VERSION);
        version = (versionString == null) ? -1 : Integer.parseInt(versionString);
        issueJql = settings.get(ISSUE_JQL);
    }

    @PreDestroy
    @Override
    public void onStop() {
        settings.set(HOST, getHost());
        settings.set(USER, getUsername());
        settings.set(PASS, getPassword());
        settings.set(VERSION, String.valueOf(version));
        settings.set(ISSUE_JQL, getIssueJql());
        settings.save();
    }

    //endregion
}

File: app/src/main/java/lt/markmerkk/utils/UserSettingsImpl.kt
package lt.markmerkk.utils

import lt.markmerkk.JiraSearchSubscriberImpl
import lt.markmerkk.listeners.WorldEvents
import lt.markmerkk.mvp.UserSettings

import javax.annotation.PostConstruct
import javax.annotation.PreDestroy

/**
 * Created by mariusmerkevicius on 12/21/15.
 * Controller for holding persistent data
 */
class UserSettingsImpl(
        private val settings: HashSettings
) : UserSettings, WorldEvents {

    override var host: String = ""
        set(value) {
            field = value
            settings.save()
        }
    override var username: String = ""
        set(value) {
            field = value
            settings.save()
        }
    override var password: String = ""
        set(value) {
            field = value
            settings.save()
        }
    override var issueJql: String = JiraSearchSubscriberImpl.DEFAULT_JQL_USER_ISSUES
        set(value) {
            field = value
            settings.save()
        }

    override var version = -1
        set(value) {
            field = value
            settings.save()
        }

    //region Getters / Setters

    override fun setCustom(key: String, value: String) {
        settings.set(key, value)
        settings.save()
    }

    override fun getCustom(key: String): String? {
        return settings.get(key)
    }

    //endregion

    //region World events

    @PostConstruct
    override fun onStart() {
        settings.load()
        host = settings.get(HOST) ?: ""
        username = settings.get(USER) ?: ""
        password = settings.get(PASS) ?: ""
        val versionString = settings.get(VERSION)
        if (versionString != null) {
            try {
                version = Integer.parseInt(versionString)
            } catch (e: NumberFormatException) {
                version = -1
            }
        } else {
            version = -1
        }
        issueJql = settings.get(ISSUE_JQL) ?: JiraSearchSubscriberImpl.DEFAULT_JQL_USER_ISSUES
    }

    @PreDestroy
    override fun onStop() {
        settings.set(HOST, host)
        settings.set(USER, username)
        settings.set(PASS, password)
        settings.set(VERSION, version.toString())
        settings.set(ISSUE_JQL, issueJql)
        settings.save()
    }

    companion object {
        val HOST = "HOST"
        val USER = "USER"
        val PASS = "PASS"
        val VERSION = "VERSION"
        val ISSUE_JQL = "ISSUE_JQL"
    }

    //endregion
}
